Manage bookings integration

MilesDev91 opened the pull request:
### What this PR does:
- This connects the front end of the Manage bookings page to the back end. All bookings are now retrieved via API calls to the request controller.
- Status can be changed via a dropdown menu for each booking. Calendar still displays dates based on upcoming bookings, but now only for accepted ones.

### Screenshots / Videos:
![image](https://user-images.githubusercontent.com/34494229/150230156-7ce59d9c-97e1-47f7-950d-98bf735147ff.png)

### Any information needed to test this feature:
- Start the server section of the app with appropriate database settings
- Make sure you have valid requests in your database. If you do not, you need:
1. At least two users, one who is the requester and one who is the sitter. You can register new users on the /register page.
2. A few requests, with one from a past date if you wish to see the past bookings section populated. You may also test without bookings to check how the page handles having no data. You may create requests through postman after logging in with a valid POST /auth/login request. As long as you get a cookie, you should then be able to create requests by calling POST at /requests with a valid body structure. Check the model to see how to structure the body.

- You will want to log in to a user who has requests as a sitter to test the status changing with the settings icon buttons.

### Any issues with the current functionality:
- The calendar does not update with the status changes until refresh, whereas the statuses themselves on the booking cards do.
- Users can click on dates with this view, which can cause confusion as to what dates their upcoming bookings are on (the styling is the same). However, they can only do this with one date.
- Their is no current distinction with whether the user is a owner or sitter, so you can navigate to the bookings page as an owner. It should not let you update status on the backend, however, even if it changes on the screen.
- There is currently no integration with profile pictures.

closes #31 

### Review thoughts:
- Bookings.tsx is getting pretty big, but I think it is still readable. Let me know if I am wrong!

MilesDev91 committed:
Create request.js

Created the request controller.

Added methods to create, edit, and get requests. Also added a method to specifically update the request status.

MilesDev91 committed:
Modified Request controller and routes

Added routes with protection for request controller. Implemented the routes in app.js

Added validation for user type query params.

Removed the patch route in the request controller. It just isn't necessary yet and will likely hurt more than help.

Cleaned up some unnecessary error codes and modified some queries in the controller.

Fixed False typo in the request model that was missed in the request model pr.

MilesDev91 committed:
Update request.js

Clarified the query param in the route comment for getUserRequests

MilesDev91 committed:
Updated request controller

Simplified one route and added a param to another.

Updated controllers to avoid directly setting to the req.body for security reasons. Enforced the sitter user type for editing requests as well.

Changed the get controller to avoid using a request body.

MilesDev91 committed:
Merge branch 'request_controller' into manage-bookings-integration

MilesDev91 committed:
Update controller and route

We pass in user id through req.user.id, so no parameter needed.

MilesDev91 committed:
Updated request controller and route

Removed passing things like user type and user id through params, and adjusted the code/routes to maintain functionality.

Removed setting paid status through request controller.

MilesDev91 committed:
Retrieving data from database

Replaced mock bookings with real bookings. Display of data currently in progress.

Renamed the Booking interface to Request to be more consistent with other parts of the application. Might add a Booking specific interface.

MilesDev91 committed:
Merge branch 'request_controller' into manage-bookings-integration

MilesDev91 committed:
API Data for Bookings

Returning data from the API for bookings. The GET method for requests now also retrieves user profiles for every request.

Changed startTime/endTime to startDate/endDate in the request interface and relevant areas.

Changed the date format to be more descriptive and readable.

Moved many of the values to useState() to trigger update on API load.

Fixed typo in the user controller.

MilesDev91 committed:
Merge branch 'main' into manage-bookings-integration

MilesDev91 committed:
Created RequestStatusButton

Encapsulated the SettingsIcon in a RequestStatusButton component. This component will display a dropdown menu when clicked that changes the request status.

Added the changeRequestStatus API helper that takes in an id and status.

Removed the userType param and query from getRequests helper.

Added an id to the Request interface for editing functionality.

Fixed calendar active class on dates with single digits.

Fixed typo in Request model

MilesDev91 committed:
Implemented status changing

The settings icon now drops a menu that changes status depending on which option you select. Added handlers to the parents to fetch changes to the database.

Added state management to BookingCard and Booking to update status on screen.

Calendar now only styles dates that have an accepted booking. Calendar currently does not update dynamically with status changes.

Fixed get request in our API helper (was missing an id param).

Now displays status on next booking. Moved the label and settings icon to the BookingCard so we can update status properly.

Status setting icon no longer displays on past bookings.

MilesDev91 committed:
Update Bookings.tsx

Removed useless imports and comments.

sundayezeilo commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -4,95 +4,79 @@ import useStyles from './useStyles';
+          const userBookings: Request[] = [];
+          for (let i = 0; i < res.requests.length; i++) {
+            userBookings[i] = {
+              // The API returns the Date object as a string, so we have to recreate it
```
Comment is unnecessary. Please remove.

sundayezeilo commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -4,95 +4,79 @@ import useStyles from './useStyles';
+            };
+          }
+          if (userBookings.length > 0 && userBookings != undefined) {
+            // The order of these operations is important
```
Remove comments to keep code clean

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,60 @@
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler")
+
+// @route POST /requests
+// @desc Create new request
+// @access Public
```
Good documentation

MilesDev91 replied:
Thanks!

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,60 @@
+  }
+});
+
+// @route PUT /requests/edit/:requestId
+// @desc edit request
+// @access Public
```
:+1:

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,60 @@
+    res.status(404);
+    throw new Error("Request doesn't exist");
+  }
+  // Until we implement cancelling bookings, owners shouldn't be able to edit bookings
```
Remove comment

sundayezeilo commented on server/controllers/request.js:
```diff
@@ -0,0 +1,60 @@
+  }
+});
+
+// @route GET /requests
+// @desc get requests of logged in user
+// @access Private
```
:+1:

sundayezeilo reviewed (commented):
I'm not sure if you linked the correct issue

MilesDev91 commented:
Fixed and linked the correct issue

MilesDev91 committed:
Cleaned up some comments

rajivtitus commented on client/src/components/BookingCard/BookingCard.tsx:
```diff
@@ -1,44 +1,81 @@
+    return formattedDate;
+  };
+
+  const statusColor = () => {
```
We can use a light-weight package like [clsx](https://www.npmjs.com/package/clsx) to conditionally apply styles directly within the MUI component

MilesDev91 replied:
Will do. That is much cleaner.

rajivtitus commented on client/src/components/BookingCard/BookingCard.tsx:
```diff
@@ -1,44 +1,81 @@
+  const handleStatusChange = (status: string) => {
+    setStatus(status);
+    booking.status = status;
+    changeRequestStatus(booking.id, status);
```
Why don't we update the snackbar here based on the success/error response from this API call

MilesDev91 replied:
Good idea.

rajivtitus commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -4,95 +4,77 @@ import useStyles from './useStyles';
-      status: 'accepted',
-    },
-  ];
+  const [mounted, setMounted] = useState(false);
```
What is the purpose of this state?

MilesDev91 replied:
To prevent an infinite loop with useEffect. I am certain there is a better way but I really couldn't figure it out at the time. If I tried to use an empty array as dependencies it threw an exhaustive deps error.

rajivtitus commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -4,95 +4,77 @@ import useStyles from './useStyles';
+      setMounted(true);
+      getRequests().then((res) => {
+        if (res) {
+          const userBookings: Request[] = [];
```
Please resort to using the built in array functions in JS unless it doesn't fit our use case

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,59 @@
+  const requests = await Request.where({ $or: [{ sitter: req.user.id }, { requester: req.user.id }] });
+  const requestProfiles = []
+  for (let i = 0; i < requests.length; i++) {
+    requestProfiles[i] = await Profile.findOne({$and: [{ $or: [{ userId: requests[i].sitter}, { userId: requests[i].requester}]}, { userId: { $ne: req.user.id }}]})
```
This approach works but I feel a simpler approach would be to add a ref to the Profile in the User model - that way we can just populate the profile for each user document 👍🏽

MilesDev91 replied:
I am confused as to how the user documents relate to this. The Request model holds a reference to the Profile id in the sitter and requester fields. The profile ids are references to the User id, so comparing userId to the request and sitter fields accomplishes the task. Could you clarify what you mean?

rajivtitus reviewed (commented):
Good work Miles! Just a few comments for you to have a look

MilesDev91 committed:
Cleaning up and Snackbar

Added messages to the Snackbar on request status change.

Used clsx to clean up some code. Refactored the formatted date to use string literals.

Streamlined the RequestStatusButton. Moved onStatusChange directly to the handlers instead of relying on state change to call it in useEffect(), especially since the state was unnecessary in the first place.

Used .push instead of assigning a value to a specific element in an array by index when populating the userBookings object in Bookings.

rajivtitus commented on client/src/components/BookingCard/BookingCard.tsx:
```diff
@@ -13,40 +15,33 @@ interface Props {
     booking.status = status;
-    changeRequestStatus(booking.id, status);
+    changeRequestStatus(booking.id, status).then((data) => {
+      console.log(data);
```
Please remove log to the console

MilesDev91 committed:
Update BookingCard.tsx

Removed a debug console log

MilesDev91 committed:
Refactored controller

Users now hold a reference to their profile on creation, creating a two way reference.

Removed petIds from the request model, as it wasn't in scope for this version of the application.

Request controller refactored to use Profile references and the isSitter value.

Fixed 'delayed' behavior in changing status. Was changing status based on the previous state.

MilesDev91 committed:
Better solution

Turns out we didn't need the two way relationship between users and profiles.

MilesDev91 committed:
Update auth.js

MilesDev91 committed:
FindOne

Was using findOne() wrong.

Bookings should now render without crashing if there is an error.

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,67 @@
+exports.getUserRequests = asyncHandler(async (req, res, next) => {
+  const profile = await Profile.findOne({'userId': req.user.id})
+  if (profile.isSitter) {
+    const requests = await Request.where('sitter', profile._id);
```
Alternatively, using a mongoose query to populate the `requester` will make things a lot more simple

Eg: `const requests = await Request.find({ sitter: profile._id }).populate('requester');`

MilesDev91 replied:
Didn't know you could do that in a query with more than one result. Thanks!

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,67 @@
+    throw new Error("Request doesn't exist");
+  }
+  const profile = await Profile.findOne({'userId': req.user.id})
+  if (profile.isSitter) {
```
The current design will allow any sitter to update other sitters' requests - I think we've addressed this in the controllers PR but I'm not sure why you reverted the validation here :)

Eg: `(request.sitter == profile._id)`

MilesDev91 replied:
I thought I was clever refactoring the code to use the new isSitter property but didn't realize the consequence. My bad

rajivtitus reviewed (approved):
Good work! Please address the comments prior to merging 👍🏽

MilesDev91 committed:
Fixed controller

Simplified retrieval of request profiles by populating the profile field on each request. Fixed front end to match new code.

Properly checked against user id for edited status so they only change their own requests.

MilesDev91 committed:
Merge branch 'main' into manage-bookings-integration

MilesDev91 committed:
Update validate.js
